### PR TITLE
Fix `Colorbar` with all zero heatmap

### PR DESCRIPTION
**Pull request by t-bltg**
# Description

Fixes https://github.com/JuliaPlots/Makie.jl/issues/1738.

With pr:
hm-mid

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] Added unit tests for new algorithms, conversion methods, etc.

### CODE REVIEW

File: src/makielayout/layoutables/colorbar.jl [hunk 1]
```diff
@@ -78,7 +78,8 @@ function layoutable(::Type{<:Colorbar}, fig_or_scene; bbox = nothing, kwargs...)
         if all(!isnothing, (limits, colorrange))
             error("Both colorrange + limits are set, please only set one, they're aliases. colorrange: $(colorrange), limits: $(limits)")
         end
-        return something(limits, colorrange, (0, 1))
+        lo, hi = lims = something(limits, colorrange, (0, 1))
+        return lo == hi ? (zero(lo), one(hi)) : lims
```
Comment (SimonDanisch) [inline src/makielayout/layoutables/colorbar.jl:82]: shouldn't this be something like `(hi - 0.5, hi + 0.5)` ?

Comment (t-bltg) [reply src/makielayout/layoutables/colorbar.jl:82]: Oh, yes I forgot about centering, sorry, will fix. I also tried to fix this after in the pipeline (in `src/makielayout/lineaxis.jl`), but without success.

Comment (t-bltg) [reply src/makielayout/layoutables/colorbar.jl:82]: Oddly, when applying `(oftype(lo, lo - .5), oftype(hi, hi + .5))` instead, I've got this

z_rng

, which I'm afraid is obviously wrong.

Comment (jkrumbiegel) [reply src/makielayout/layoutables/colorbar.jl:82]: the heatmap also needs the colorrange

Comment (jkrumbiegel) [reply src/makielayout/layoutables/colorbar.jl:82]: what you did is build a colorbar that takes its colormap from the heatmap but sets different limits, without synching those back to the heatmap.

Comment (t-bltg) [reply src/makielayout/layoutables/colorbar.jl]: Thanks for pointing that out. Here is the new output:
hm-mid

**Review by SimonDanisch — APPROVED** [review]
Sounds good to me! @jkrumbiegel  any concerns?

**Comment by jkrumbiegel** [comment]
Maybe the test should check what the limit values in the colorbar are? Not sure if that's accessible actually. And what happens when the heatmap observable changes so there is now a nonempty range, will the colormap adjust correctly? Maybe that should be included in the test. Also, the 0.5 offset might lead to problems for really large float32 singular values, maybe a factor off 1.5 or so could work? Unless zero

**Comment by t-bltg** [comment]
> Maybe the test should check what the limit values in the colorbar are

I've added test for `colorrange` and `limits`.

> And what happens when the heatmap observable changes so there is now a nonempty range

I've also added a test changing either `hm.colorrange` or `cb.limits`.

> Also, the 0.5 offset might lead to problems for really large float32 singular values

As long as you have `hi - lo > eps(Float32)`, I think that should suffice to avoid the initial error `ERROR: ArgumentError: range step cannot be zero`.

**Comment by t-bltg** [comment]
I've found a way to change the `colorrange` upstream in the pipeline, and updated the PR.
This is cleaner IMO, please comment.

**Review by jkrumbiegel — APPROVED** [review]

**Comment by SimonDanisch** [comment]
Thank you! That looks like a fair solution!

**Pull request merged**